modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1245 @@
+
+      :arg parSafe: If `true`, this unrolledLinkedList will use parallel safe operations.
+      :type parSafe: `param bool`
+    */
+    proc init(other: [?d] ?t, param parSafe=false, nodeCapacity: int = 32) {
+      _checkType(t);
+      if !isCopyableType(t) then
+        compilerError("Cannot construct unrolledLinkedList from array with element " +
+                      "type that cannot be copied");
+
+      this.eltType = t;
+      this.parSafe = parSafe;
+      this.nodeCapacity = nodeCapacity;
+      this.complete();
+      _commonInitFromIterable(other);
+    }
+
+    /*
+      Initializes a unrolledLinkedList containing elements that are copy initialized from
+      the elements contained in a list.
+
+      :arg other: The list to initialize from.
+    */
+    proc init=(other: unrolledLinkedList(this.type.eltType, ?p)) {
+      if !isCopyableType(this.type.eltType) then
+        compilerError("Cannot construct unrolledLinkedList from list with " +
+        "element type that cannot be copied");
+      _checkType(this.type.eltType);
+      _checkNodeCapacity(other.nodeCapacity);
+
+      this.eltType = this.type.eltType;
+      this.parSafe = this.type.parSafe;
+      this.nodeCapacity = other.nodeCapacity;
+      this.complete();
+      _commonInitFromIterable(other);
+    }
+
+    pragma "no doc"
+    proc _commonInitFromIterable(iterable) {
+      for x in iterable do
+        append(x);
+    }
+
+    /*
+    pragma "no doc"
+    proc _firstTimeInitializeArrays() {
+      _sanity(_head == nil);
+      _sanity(_tail == nil);
+      _sanity(_size == 0);
+      _head = new unmanaged _linkedNode(eltType);
+      _tail = _head;
+    }
+    */
+
+    //
+    // A helper function for getting a reference to a unrolledLinkedList element.
+    //
+    pragma "no doc"
+    inline proc const ref _getRef(idx: int) ref {
+      _sanity(idx >= 0 && idx < _size);
+      _sanity(_head != nil);
+      var cur = _head, pos = 0;
+      while cur != nil {
+        if idx < pos + cur!.size {
+          ref result = cur!.data[idx - pos];
+          return result;
+        }
+        else {
+          pos += cur!.size;
+        }
+        cur = cur!.next;
+      }
+      halt("unrolledLinkedList out of range");
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    pragma "no doc"
+    inline proc const _withinBounds(idx: int): bool {
+      return (idx >= 0 && idx < _size);
+    }
+
+    /*
+     Split a node from the given node p
+     by moving half of the content of p into the new one
+    */
+    pragma "no doc"
+    proc _split(p: unmanaged _linkedNode(eltType)) {
+      _sanity(p.size == nodeCapacity);
+      var node = new unmanaged _linkedNode(eltType, nodeCapacity);
+      var start = p.size/2;
+      for i in start..p.size-1 {
+        node.append(p.data[i]);
+      }
```

Comment:
... even though those entries in `node` are the only ones that exist according to the public interface of `unrolledLinkedList`.

Reply:
Sounds reasonable. How can I release these references then?

Reply:
In the case of ULL, a shared object must be `nilable shared` because it has to be default initializable. So the solution is to  set it to `nil`. Do you think using`_move` here can solve this? I don't know about what happens to the place of the element moved.

Reply:
Sounds like you need a `p.data[i].clear` after you append it to the new node.

https://chapel-lang.org/docs/master/builtins/SharedObject.html